src/librustc/hir/def_id.rs

```diff
@@ -152,8 +161,21 @@ impl DefId {
     }
 }
 
-impl rustc_serialize::UseSpecializedEncodable for DefId {}
-impl rustc_serialize::UseSpecializedDecodable for DefId {}
+impl rustc_serialize::UseSpecializedEncodable for DefId {
+    fn default_encode<S: Encoder>(&self, s: &mut S) -> Result<(), S::Error> {
+        let krate = u64::from(self.krate.as_u32());
+        let index = u64::from(self.index.as_u32());
+        s.emit_u64((krate<<32) | index)
```

Comment:
Similar to `CrateNum`, one should include a `DefId` to `DefPath` mapping in the serialized data. However, in this PR, `DefId` isn't used by the linker so that the mapping can be safely omitted.